File: chapter3/019.rb
n = gets.chomp.to_i
arr = gets.chomp.split(" ").map(&:to_i)

colors = {
  red: 0,
  yellow: 0,
  blue: 0,
}

arr.each do |i|
  case i
  when 1
    colors[:red] += 1
  when 2
    colors[:yellow] += 1
  when 3
    colors[:blue] += 1
  end
end

red = (colors[:red] * (colors[:red] - 1)) / 2
yellow = (colors[:yellow] * (colors[:yellow] - 1)) / 2
blue = (colors[:blue] * (colors[:blue] - 1)) / 2

puts [red, yellow, blue].sum

File: chapter3/019.exs
defmodule Main do
  def main do
    _n = Acl.Input.get_n()

    Acl.Input.get_n_list()
    |> solve()
    |> IO.puts()
  end

  defp solve(n_list) do
    colors =
      Enum.reduce(n_list, %{1 => 0, 2 => 0, 3 => 0}, fn a, acc ->
        Map.update(acc, a, 1, &(&1 + 1))
      end)

    red = div(colors[1] * (colors[1] - 1), 2)
    yellow = div(colors[2] * (colors[2] - 1), 2)
    blue = div(colors[3] * (colors[3] - 1), 2)

    red + yellow + blue
  end
end

defmodule Acl do
  defmodule Input do
    def get_n, do: IO.read(:line) |> String.trim() |> String.to_integer()

    def get_s, do: IO.read(:line) |> String.trim()

    def get_n_list,
      do: IO.read(:line) |> String.trim() |> String.split(" ") |> Enum.map(&String.to_integer/1)

    def get_s_list,
      do: IO.read(:line) |> String.trim() |> String.split(" ")

    def get_s_split, do: IO.read(:line) |> String.trim() |> String.codepoints()
  end
end

Main.main()
